Support all five EdDSA instances from RFC 8032

jamuir opened the pull request:
Fixes #6277

Adds support for all five EdDSA instances from RFC 8032: **Ed25519, Ed25519ctx, Ed25519ph, Ed448, Ed448ph**.

Only Ed25519 was already fully supported (via the EVP APIs)

All instances, except for Ed25519, allow context strings as input.  Context strings can now be passed via an OSSL_PARAM.

The desired EdDSA instance can also be specified via an OSSL_PARAM.

<!--
Thank you for your pull request. Please review these requirements:

Contributors guide: https://github.com/openssl/openssl/blob/master/CONTRIBUTING.md

Other than that, provide a description above this comment if there isn't one already

If this fixes a GitHub issue, make sure to have a line saying 'Fixes #XXXX' (without quotes) in the commit message.
-->

##### Checklist
- [ y ] documentation is added or updated
- [ y ] tests are added or updated

jamuir commented:
@paulidale: could you take a look at the changes so far?

slontis commented on dev/testing/6277/EdDSA.c:
```diff
@@ -0,0 +1,210 @@
+#include <stdio.h>
```
This directory needs to disappear, and be replaced by files sitting in the test folder.
Search for ecx, ed and X25519, X448 in the test folder..

Quite a few of the algorithm tests are driven by test/evp_test.c...
(See l/test/recipes/30-test_evp_data/evppkey_ecx.txt)

jamuir replied:
this is temporary.  I just needed somewhere to track my dev test.

I will eventually move it to someplace sensible.

slontis commented on include/crypto/ecx.h:
```diff
@@ -33,15 +33,15 @@
 #  define X25519_SECURITY_BITS  128
 
-#  define X448_BITS             448
+#  define X448_BITS             448  // ?? but the group order is 446 bits
```
// Is not good for some compiler...
@mattcaswell might be the best person to answer your question..

mattcaswell replied:
I *think* this is the key size (56 bytes == 448)

jamuir replied:
okay.   the values for `X448_BITS`, `ED25519_BITS`, `ED448_BITS` make sense if they are meant to be the bit-length of the secrets/private keys.  But, the value for `X25519_BITS` is 253 -- that does not fit that pattern.  A raw X25519 secret is 32 bytes.

In RFC 7748, there is the following paragraph:

>  Scalars are assumed to be randomly generated bytes.  For X25519,
in order to decode 32 random bytes as an integer scalar, set the
three least significant bits of the first byte and the most
significant bit of the last to zero, set the second most
significant bit of the last byte to 1 and, finally, decode as
little-endian.  This means that the resulting integer is of the
form 2^254 plus eight times a value between 0 and 2^251 -
1 (inclusive).  Likewise, for X448, set the two least significant
bits of the first byte to 0, and the most significant bit of the
last byte to 1.  This means that the resulting integer is of the
form 2^447 plus four times a value between 0 and 2^445 -
1 (inclusive).

Maybe someone set `X25519_BITS` to 253 because of this paragraph.

in any case, maybe this discussion needs to be moved elsewhere.  I can open an issue for it if you think it needs one.

jamuir replied:
seems related to #19070

mattcaswell replied:
Yeah...looks like there is some inconsistency here.

slontis commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -43,6 +43,30 @@ static int s390x_ed448_digestverify(const ECX_KEY *edkey,
+       official NID assignments are recorded inside obj_mac.h.
+       we could also just use an enum instead. */
+
+#define NID_Ed25519_INSTANCE    NID_ED25519
```
There doesnt seem to be a good reason to use NID here.. The pattern used in other files is to use a table lookup.. e.g.

````
#define KEM_OP_UNDEFINED   -1
#define KEM_OP_RSASVE       0

static const OSSL_ITEM rsakem_opname_id_map[] = {
    { KEM_OP_RSASVE, OSSL_KEM_PARAM_OPERATION_RSASVE },
};

static int name2id(const char *name, const OSSL_ITEM *map, size_t sz)
{
    size_t i;

    if (name == NULL)
        return -1;

    for (i = 0; i < sz; ++i) {
        if (OPENSSL_strcasecmp(map[i].ptr, name) == 0)
            return map[i].id;
    }
    return -1;
}

static int rsakem_opname2id(const char *name)
{
    return name2id(name, rsakem_opname_id_map, OSSL_NELEM(rsakem_opname_id_map));
}
````

jamuir replied:
okay.  I will refactor.

slontis commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -326,6 +465,76 @@ static const OSSL_PARAM *eddsa_gettable_ctx_params(ossl_unused void *vpeddsactx,
+    return settable_ctx_params;
+}
+
 const OSSL_DISPATCH ossl_ed25519_signature_functions[] = {
```
I don't think we need separate dispatch tables here since the algorithms are very similar.. Using a mode via a settable seems like the right approach,

jamuir replied:
thanks.  so things are good as I've put them (i.e. you are not suggesting I change something here).

mattcaswell commented on crypto/ec/curve25519.c:
```diff
@@ -5576,6 +5576,216 @@ ossl_ed25519_verify(const uint8_t *message, size_t message_len,
+    };
+#else
+    const char dom_s[] = "SigEd25519 no Ed25519 collisions";
+#endif
```
I think we now tend to prefer just always specifying it in hex, e.g. see:

https://github.com/openssl/openssl/blob/3c4e250eafd789bb8ff905cd2ec9dddc7e1656d2/ssl/tls13_enc.c#L416-L431

jamuir replied:
ack.  I can change that.  Note that I was following the example inside `crypto/ec/curve448/eddsa.c`.

Do you want me to change `crypto/ec/curve448/eddsa.c` as well?

mattcaswell replied:
> Do you want me to change crypto/ec/curve448/eddsa.c as well?

That would be a nice-to-have.

mattcaswell commented on crypto/ec/curve25519.c:
```diff
@@ -5576,6 +5576,216 @@ ossl_ed25519_verify(const uint8_t *message, size_t message_len,
 }
 
+/*
+   JM: based on hash_init_with_dom() from eddsa.c (used with Ed448).
```
Remove the JM prefix

mattcaswell reviewed (changes requested):
Not a proper review. I just scanned through it and commented on a few things. I agree with most of what @slontis says.

jamuir commented:
@slontis @mattcaswell :  thank-you for taking a look!

jamuir commented on crypto/ec/curve25519.c:
```diff
@@ -5576,6 +5576,216 @@ ossl_ed25519_verify(const uint8_t *message, size_t message_len,
+
+int
+ossl_ed25519_dom2_sign(uint8_t *out_sig, const uint8_t *tbs, size_t tbs_len,
+                       const uint8_t public_key[32], const uint8_t private_key[32],
```
Initially, I was concerned about breaking the existing implementation -- it felt safer to implement a separate function.  And I also thought I might need to implement HashEdDSA as a completely separate signature scheme.

I will move the new `dom2_sign/verify` functionality into the existing functions.

beldmit commented:
JFYI.

https://hdevalence.ca/blog/2020-10-04-its-25519am

jamuir commented:
@beldmit :  was there anything in particular from that blog that you wanted to point out?

Note that I do not intend to modify any curve arithmetic.  In particular, for edwards25519, the strict verification equation is used currently and I will not change that in this PR.

jamuir commented:
All five EdDSA instances are working now.  The HashEdDSA instances are supported only in "one-shot" mode.

This PR is close to being ready, but I still need to make some documentation updates and relocate my dev test.

jamuir commented:
@slontis @mattcaswell :  this PR is ready (i.e. it is no longer work-in-progress).

I believe I have addressed all your previous comments.

Please let me know if you have any additional concerns.

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+            NULL, testctx,
+            NULL, pkey_sk,
+            params);
+    if (rc==0) {
```
nit: spaces around `==`

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+            &sig_length,
+            rfc8032_test_vectors[i].msg,
+            rfc8032_test_vectors[i].msg_length);
+    if (rc==0) {
```
nit: spaces around `==`

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+            NULL, testctx,
+            NULL, pkey_pk,
+            params);
+    if (rc==0) {
```
nit: spaces around `==`

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+            sig_length,
+            rfc8032_test_vectors[i].msg,
+            rfc8032_test_vectors[i].msg_length);
+    if (rc==0) {
```
nit: spaces around `==`

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+        goto err;
+    }
+
+    printf("passed\n");
```
Remove printf statements

hlandau commented on test/evp_extra_test.c:
```diff
@@ -4717,6 +4721,161 @@ static int test_ecx_not_private_key(int tst)
+            rfc8032_test_vectors[i].msg,
+            rfc8032_test_vectors[i].msg_length);
+    if (rc==0) {
+        printf("failed (%d)\n", __LINE__);
```
Remove printf and use TEST_error or similar here

jamuir replied:
printf's are gone and nits picked.  thx!

jamuir commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -29,17 +73,24 @@ EVP_PKEY_CTX_get_params().
 EVP_DigestSign() and EVP_DigestVerify() functions.
 
+The HashEdDSA instances do not yet support the streaming mechanisms
+(so the one-shot functions must be used with HashEdDSA as well).
```
Yes, the reason is that adding support for the streaming mechanisms is quite a bit more work.  I did not want to make that time investment yet.

But I would be willing to work on it later (in a separate PR).

jamuir commented on crypto/ec/curve25519.c:
```diff
@@ -5434,9 +5434,50 @@ static void sc_muladd(uint8_t *s, const uint8_t *a, const uint8_t *b,
+    dom[0] = (uint8_t)(phflag >= 1 ? 1 : 0);
+    dom[1] = (uint8_t)context_len;
+
+    if (!EVP_DigestUpdate(hash_ctx, dom_s, strlen(dom_s))
```
I've changed this in both places using `sizeof(dom_s)-1`.

jamuir commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,16 +10,60 @@ Ed448
+
+=item "instance" (B<OSSL_SIGNATURE_PARAM_INSTANCE>) <utf8 string>
+
+One of the five strings "Ed25519", "Ed25519ctx", "Ed25519ph", "Ed448", "Ed448ph".
```
I've updated the doc as suggested.

jamuir commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -167,29 +258,75 @@ int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
+        if (peddsactx->dom2_flag
+                || peddsactx->context_string_flag
+                || peddsactx->context_string_len > 0) {
+            /* s390x_ed25519_digestsign() does not yet support dom2 or context-strings */
```
I considered that, but I didn't want users to puzzle over a mysterious slow down in EdDSA performance on s390x (e.g. "why is `Ed25519ph` slower than `Ed25519` on s390x?").  Having the call fail provides more immediate feedback to the user (and possibly provides motivation for someone to get this working on s390x).

I don't have a strong opinion here.  Maybe it is fine to do as you suggest.

I guess I can add a new comment in the code that devs can point to later on when someone notices the performance drop.

jamuir replied:
I've just pushed a new revision with this change.

hlandau replied:
I don't think the proposed behaviour works to give 'immediate feedback'. If someone ships an application which uses these algorithms, it will suddenly break when porting to s390x. If we support an algorithm on one platform we should be supporting it on all platforms, regardless of the performance we can offer.

jamuir commented:
@hlandau :  thanks for your comments.

I believe I have addressed everything you raised.

Please let me know if you have other suggestions.

slontis commented on crypto/ec/curve25519.c:
```diff
@@ -5448,6 +5489,17 @@ ossl_ed25519_sign(uint8_t *out_sig, const uint8_t *message, size_t message_len,
+
+    /* if csflag is set, then a non-empty context-string is required */
+    if (csflag && context_len == 0)
+        goto err;
```
the first 2 gotos could be changed to return 0 if you do the fetch and new later..

jamuir replied:
Agreed, but I don't think that change offers much benefit.  I prefer to leave the code as is.

slontis replied:
See chapter 7 of https://www.openssl.org/policies/technical/coding-style.html

slontis commented on crypto/ec/curve25519.c:
```diff
@@ -5434,9 +5434,50 @@ static void sc_muladd(uint8_t *s, const uint8_t *a, const uint8_t *b,
+            "\x6c\x6c\x69\x73\x69\x6f\x6e\x73";
+    uint8_t dom[2];
+
+    if (sha512 == NULL || hash_ctx == NULL)
```
These checks are done by the caller already..

jamuir replied:
ok.  I will remove the null checks.

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+RFC 8032 describes five EdDSA instances: Ed25519, Ed25519ctx,
+Ed25519ph, Ed448, Ed448ph.
+
+The instances Ed25519, Ed25519ctx, Ed448 are referred to as B<PureEdDSA>
```
.. and Ed448

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+schemes.  For these three instances, the sign and verify procedures
+require access to the complete message (not a digest of the message).
+
+The instances Ed25519ph, Ed448ph are referred to as B<HashEdDSA>
```
and

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+Ed448ph, the hash function is SHAKE256 with an output length of 512
+bits.
+
+The instances Ed25519ctx, Ed25519ph, Ed448, Ed448ph allow an optional
```
and

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+
+=over 4
+
+=item * "instance" (B<OSSL_SIGNATURE_PARAM_INSTANCE>) <utf8 string>
```
Is there a reason for the "instance" name, type or mode maybe?

jamuir replied:
Yes, the reason is consistency with RFC 8032.

Ed25519, Ed25519ctx, Ed25519ph, Ed448, Ed448ph are all described as EdDSA instances in RFC 8032.

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+
+=item * "instance" (B<OSSL_SIGNATURE_PARAM_INSTANCE>) <utf8 string>
+
+One of the five strings "Ed25519", "Ed25519ctx", "Ed25519ph", "Ed448", "Ed448ph".
```
or

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+
+One of the five strings "Ed25519", "Ed25519ctx", "Ed25519ph", "Ed448", "Ed448ph".
+
+"Ed25519", "Ed25519ctx", "Ed25519ph" are valid only for an Ed25519 EVP_PKEY.
```
and

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -64,16 +119,22 @@ specified, then both Ed25519 and Ed448 are benchmarked.
 
-        EVP_DigestSignInit(md_ctx, NULL, NULL, NULL, ed_key);
-        /* Calculate the requires size for the signature by passing a NULL buffer */
+        const OSSL_PARAM params[] = {
```
Maybe add a comment saying params can be NULL if ...

jamuir replied:
good idea.  I will do that.

slontis commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -166,30 +257,75 @@ int ed25519_digest_sign(void *vpeddsactx, unsigned char *sigret,
+static int ed448_shake256(OSSL_LIB_CTX *libctx,
+                          const char *propq,
+                          const uint8_t *in, size_t inlen,
+                          uint8_t *out, size_t outlen )
```
outlen)

jamuir replied:
thx.   I will remove the space.

slontis commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -326,6 +509,85 @@ static const OSSL_PARAM *eddsa_gettable_ctx_params(ossl_unused void *vpeddsactx,
+        if (!OSSL_PARAM_get_utf8_string(p, &pinstance_name, sizeof(instance_name)))
+            return 0;
+
+        if (OPENSSL_strcasecmp(pinstance_name, SN_Ed25519) == 0) {
```
You could have a table for these.. and then just loop over the table in a function.

jamuir replied:
if there were a few more params, then I would do as you suggest.

slontis commented on providers/implementations/signature/eddsa_sig.c:
```diff
@@ -341,9 +603,13 @@ const OSSL_DISPATCH ossl_ed25519_signature_functions[] = {
     { 0, NULL }
 };
 
+/* TODO: add streaming APIs for Ed448ph */
```
Take the todo comment out.. We end up with lots of TODO's everywhere.. Create a issue if you need a reminder.

jamuir replied:
okay.  I will remove the "TODO" comments.

slontis commented on test/rfc8032-test-vectors.h:
```diff
@@ -0,0 +1,779 @@
+#ifndef RFC8032_TEST_VECTORS_H
```
Why didnt you use evp_test.c?

jamuir replied:
I'm not sure I understand.

Are you asking why I chose to add the rfc8032 test vectors in a new header file rather than just add them directly inside `evp_extra_test.c` ?

It just didn't occur to me to do that.  There is quite a lot of test vector data (the header file is over  700 lines).  Keeping them in a separate header seemed reasonable.

slontis replied:
This question was - why did you not use evp_test.c.
evp_test.c loads files from test/recipes/30-test_evp_data/*.txt. 
A lot of algorithms are tested this way. i.e Instead of being included from a header does it make sense to add this to a txt file and use evp_test.c

t8m replied:
Would it be too much to ask to convert these testcases to evp_test data testcases?

It would require modification of evp_test.c to pass the algorithm variant as params though.

jamuir replied:
Hi Tomas.  Refactoring the test and reformatting all the test vector data again sounds really fun :wink: 

Seriously though, I don't mind changing it, but I would prefer to do that in a separate PR.

jamuir commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+=head2 EdDSA Instances
+
+RFC 8032 describes five EdDSA instances: Ed25519, Ed25519ctx,
+Ed25519ph, Ed448, Ed448ph.
```
I don't think adding "and" is an improvement (note that I left it out deliberately).

I prefer to leave this as is (and similarly for the other "and" and "or" suggestions below).

slontis replied:
This is not consistent with what has been done in other newer pod files .

jamuir replied:
Can you give a pointer to those pod files?

slontis replied:
search for and in EVP_PKEY*.pod, EVP_KDF*.pod

jamuir commented:
Strangely, the test `buildbot/master:unix-macos11-x86_64` failed:

```
15-test_rsaoaep.t .................. ok
15-test_rsapss.t ................... ok
15-test_sha.t ...................... ok
20-test_app.t ...................... ok
20-test_cli_fips.t ................. skipped: Test only supported in a fips build with security checks
20-test_dgst.t ..................... ok
20-test_dhparam.t .................. ok

command timed out: 1200 seconds without output running [b'make', b'test'], attempting to kill
process killed by signal 9
program finished with exit code -1
elapsedTime=1952.749765
```

https://ci.buildbot.openssl.org/#/builders/16/builds/1736

I don't understand why this particular platform failed.

is this a test infrastructure issue?

slontis commented:
buildbot runners seem to be unreliable currently.

jamuir commented:
@mattcaswell @hlandau:  thanks for the reviews you left on 18 Nov and 30 Nov.  Each of you requested changes, but I believe my updates have addressed all your comments.  Would you be able to update your reviews?

slontis reviewed (commented):
Here is example converted  test data for test vector #21 (using your .h file) that I appended to test/recipes/30-test_evp_data/evppkey_ecx.txt to test this works..

So the conversion is not hard and should be part of this PR.

````
# Test Vector 21
# Raw versions of the Ed448ph keys

PrivateKeyRaw=ED448PH-21-Raw:ED448:833fe62409237b9d62ec77587520911e9a759cec1d19755b7da901b96dca3d42ef7822e0d5104127dc05d6dbefde69e3ab2cec7c867c6e2c49

PublicKeyRaw=ED448PH-21-PUBLIC-Raw:ED448:259b71c19f83ef77a7abd26524cbdb3161b590a48f7d17de3ee0ba9c52beb743c09428a131d6b1b57303d90d8132c276d5ed3d5d01c0f53880

PrivPubKeyPair = ED448PH-21-Raw:ED448PH-21-PUBLIC-Raw

OneShotDigestSign = NULL
Key = ED448PH-21-Raw
Input = 616263
Ctrl = instance:Ed448ph
Ctrl = hexcontext-string:666f6f
Output = c32299d46ec8ff02b54540982814dce9a05812f81962b649d528095916a2aa481065b1580423ef927ecf0af5888f90da0f6a9a85ad5dc3f280d91224ba9911a3653d00e484e2ce232521481c8658df304bb7745a73514cdb9bf3e15784ab71284f8d0704a608c54a6b62d97beb511d132100
````

slontis commented:
needs a rebase also.

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -10,36 +10,91 @@ Ed448
+If the instance name is not specified, then the default "Ed25519" or
+"Ed448" is used.
+
+If a context-string is not specified, then an empty context-string is
```
I dont think you describe this case..
    /* if csflag is set, then a non-empty context-string is required */

jamuir replied:
Good point.  I've added a sentence explaining that a nonempty context string is required with Ed255ctx.

jamuir commented:
> Here is example converted test data for test vector https://github.com/openssl/openssl/pull/21 (using your .h file) that I appended to test/recipes/30-test_evp_data/evppkey_ecx.txt to test this works.

Thanks for providing this example.  I will give it a try.

jamuir commented:
I've added all 21 test vectors to `evppkey_ecx.txt` and confirmed that they are being exercised.  I've dropped the file `rfc8032-test-vectors.h`, as well as my changes to `evp_extra_test.c`.

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -64,16 +120,24 @@ specified, then both Ed25519 and Ed448 are benchmarked.
-        /* Calculate the requires size for the signature by passing a NULL buffer */
+        const OSSL_PARAM params[] = {
+            OSSL_PARAM_utf8_string ("instance", "Ed25519ctx", 10),
+            OSSL_PARAM_octet_string("context-string", (unsigned char *)"A protocol defined context string", 24),
```
You might want to assign this to a string first... The length of 24 doesnt seem to match this.

jamuir replied:
hmm... not sure why I had 24 there rather than 33.  I've fixed it now.

slontis commented on doc/man7/EVP_SIGNATURE-ED25519.pod:
```diff
@@ -64,16 +120,24 @@ specified, then both Ed25519 and Ed448 are benchmarked.
-        EVP_DigestSignInit(md_ctx, NULL, NULL, NULL, ed_key);
-        /* Calculate the requires size for the signature by passing a NULL buffer */
+        const OSSL_PARAM params[] = {
+            OSSL_PARAM_utf8_string ("instance", "Ed25519ctx", 10),
```
you can specify 0 instead of 10 here..

jamuir replied:
cool.  I didn't know that.

I think I will keep it as "10" so people reading the doc don't have to jump to OSSL_PARAM_int.html to figure out why "0" is okay.

slontis commented:
You will need to add
FIPSversion = >=3.2.0 to each of the new tests in the .txt file..
We run the tests in master against older fips providers, so this new functionality needs to be skipped when this happens.

jamuir commented:
@slontis :  thanks for applying the "tests:present" label and the detailed comments you provided previously.

I am not sure what else I can do to move this PR forward.

Would you be willing to give an approval?

slontis commented on crypto/ec/curve25519.c:
```diff
@@ -5460,20 +5509,20 @@ ossl_ed25519_sign(uint8_t *out_sig, const uint8_t *message, size_t message_len,
     ge_p3_tobytes(out_sig, &R);
 
-    if (!EVP_DigestInit_ex(hash_ctx, sha512, NULL)
+    if (!hash_init_with_dom(hash_ctx, sha512, dom2flag, phflag, context, context_len)
```
I am not sure it would be any quicker - but you could potentially use EVP_MD_CTX_dup() here

jamuir replied:
I considered using dup() at one point, but then convinced myself to do it the boring way :-)

Scalar multiplication dominates the cost of signature generation, so even if dup() is a little quicker, then it will hardly be noticeable.

slontis commented on crypto/ec/curve25519.c:
```diff
@@ -5570,6 +5632,14 @@ ossl_ed25519_verify(const uint8_t *message, size_t message_len,
 
     res = CRYPTO_memcmp(rcheck, r, sizeof(rcheck)) == 0;
+
+    /* note that we have used the strict verification equation here.
```
The following notation is used elsewhere..
````
/*
 * Blah blah
 * Blah blah2
 */
 ````

jamuir replied:
fixed.

slontis reviewed (approved):
Approved subject to NIT being fixed.

slontis commented:
@hlandau will need to reapprove,

jamuir commented:
thanks, slontis, for approving these changes.

hlandau's previous approval appears to still be valid.

@mattcaswell :  do you have some time to revisit this PR?  You requested changes on 18 Nov 2022.  I believe all your comments have been addressed.

openssl-machine commented:
This pull request is ready to merge

hlandau commented:
Merged to master. Thank you.
